src/scripts/main.js

```diff
@@ -1,3 +1,344 @@
+  } else if (e.code === moveUp) {
+    slideUp();
+  } else if (e.code === moveDown) {
+    slideDown();
+  }
+
+  document.querySelector('.game-score').innerText = score;
+
+  if (gameIsOver()) {
+    loseMessage.classList.remove('hidden');
+  }
+
+  if (isWinner()) {
+    winMessage.classList.remove('hidden');
+  }
+});
+
+startButton.addEventListener('click', () => {
+  if (!startButton.classList.contains('start')) {
+    startButton.classList.remove('restart');
+    startButton.classList.add('start');
+    startButton.innerText = 'Start';
+    gameStarted = false;
+    resetGame();
+  } else {
+    startButton.classList.remove('start');
+    startButton.classList.add('restart');
+    startButton.innerHTML = 'Restart';
+    startMessage.classList.add('hidden');
+    gameStarted = true;
+    addRandomCell();
+    addRandomCell();
+  }
+});
+
+function resetGame() {
+  score = 0;
+  document.querySelector('.game-score').innerText = score;
+  winMessage.classList.add('hidden');
+  loseMessage.classList.add('hidden');
+  startMessage.classList.remove('hidden');
+
+  board = [
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+  ];
+
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      const fieldCell = document.getElementById(r.toString() + '-' + c.toString());
+
+      updateFieldCell(fieldCell, 0);
+    }
+  }
+}
+
+function hasEmptyFieldCell() {
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      if (board[r][c] === 0) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+function addRandomCell() {
+  if (!hasEmptyFieldCell) {
+    return;
+  }
+
+  while (true) {
+    const randomRowIndex = Math.floor(Math.random() * rows);
+    const randomColumnIndex = Math.floor(Math.random() * columns);
+
+    if (board[randomRowIndex][randomColumnIndex] === 0) {
+      const randomFieldCell = document.getElementById(randomRowIndex.toString() + '-' + randomColumnIndex.toString());
+      const randomCellValue = Math.random() >= 0.1 ? 2 : 4;
+
+      board[randomRowIndex][randomColumnIndex] = randomCellValue;
+      randomFieldCell.innerText = randomCellValue;
+      randomFieldCell.classList.add(`field-cell--${randomCellValue}`);
+      break;
+    }
+  }
+}
+
+function arraysAreEqual(arr1, arr2) {
+  if (arr1.length !== arr2.length) {
+    return false;
+  }
+
+  for (let i = 0; i < arr1.length; i++) {
+    if (arr1[i] !== arr2[i]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function clearZeroes(row) {
+  return row.filter(num => num !== 0);
+}
+
+function slide(row) {
+  let newRow = clearZeroes(row);
+
+  for (let i = 0; i < newRow.length - 1; i++) {
+    if (newRow[i] === newRow[i + 1]) {
+      newRow[i] *= 2;
+      newRow[i + 1] = 0;
+      score += newRow[i];
+    }
+  }
+
+  newRow = clearZeroes(newRow);
+
+  while (newRow.length < columns) {
+    newRow.push(0);
+  }
+
+  return newRow;
+}
+
+function slideLeft() {
```

Comment:
seems like slideLeft and slideRight differ only by reversed array, let's try to join these two functions into one for not writing the same code twice